rfd/0xxx-tls-routing-behind-lb.md

```diff
@@ -0,0 +1,319 @@
+---
+authors: STeve Huang (xin.huang@goteleport.com)
+state: draft
+---
+
+# RFD XXX - TLS Routing behind load balancers
+
+## Required Approvals
+
+* Engineering: @smallinsky && @r0mant
+* Product: @xin || @klizhentas
+* Security: @reedloden
+
+## What
+
+This RFD details how to support [TLS
+Routing](https://github.com/gravitational/teleport/blob/master/rfd/0039-sni-alpn-teleport-proxy-routing.md) behind load
+balancers.
+
+## Why
+
+Allows simple single-port Teleport deployment when Telport needs to sit behind a load balancer.
+
+## Details
+
+### The challenge
+
+Here is a typical setup that we want to enable by this RFD:
+```
+ ┌───┐
+ │tsh│
+ └┬──┘
+  │
+  │SSH,DB,Kube,App,etc.
+  │
+ ┌▼────────────┐
+ │Load Balancer◄──┐
+ └┬────────────┘  │Reversse tunnel,
+  │               │Join,
+  │Web port       │etc.
+  │               │
+ ┌▼───────┐     ┌─┴──────┐
+ │Teleport│     │Teleport│
+ │Proxy(s)│     │Agent(s)│
+ └────────┘     └────────┘
+```
+
+The "load balancer" here could be a cloud-hosted service like AWS ALB, a Kubernetes ingress controller, or any other
+similar network solutions.
+
+TLS Routing already combines all proxy ports into a single port. However, most load balancers do NOT support passthrough
+of custom ALPN, SNI, or mTLS, which are required by TLS Routing, when the load balancers are configured to terminate
+TLS.
+
+### The "connection upgrade"
+
+Borrowed from the "WebSocket" design, the Teleport client can make a connection upgrade through a web API on the
+Teleport Proxy:
```

Comment:
I would include @rosstimothy on this review to make sure this upgrade does not erase all the connection latency improvements he's made in the past year.

Reply:
Will try to add some benchmarks later this week to the performance section at the bottom. The extra latency is certainly noticeable in my previous POC.

Reply:
Some preliminary data collected based on `tc.GetClusterAlerts` from `tsh login`.
https://github.com/gravitational/teleport/blob/702a3ebd32390a4660a3631150c280647423d6ec/lib/client/api.go#L2069-L2085

Two conn upgrades are done, one for the proxy SSH client (`connectToProxy`), and one for the auth client for the GRPC call.

Without conn upgrade it took 450ms, and with conn upgrade it took 570ms. My ping to Proxy is about 30ms. So about a 27% increase in latency for this particular case. ~ 2 RTT per upgrade.

Sad news in my POC, 9 conn upgrades are done per `tsh login`.

Reply:
Updated the performance section a little bit. The original plan was to implement the basic feature first and then look at possible performance improvement separately, but love to hear your thoughts on that.